Migrate scripts to mbot Diagnostics 2.0

abrahambotros opened the pull request:
For each monitor, internal stats are unchanged, but publishing is
done through mbot Diagnostics 2.0.

Required for https://github.com/MarbleInc/mbot/pull/3196.

**Please do not squash merge.**

Commit:
Migrate scripts to mbot Diagnostics 2.0

For each monitor, internal stats are unchanged, but publishing is
done through mbot Diagnostics 2.0.

zacwitte commented on bin/cpu_monitor.py:
```diff
@@ -91,7 +94,18 @@ def update_status_stale(stat, last_update_time):
     def __init__(self, hostname, namespace, diag_hostname):
-        self._diag_pub = rospy.Publisher('/diagnostics', DiagnosticArray, queue_size = 100)
+        self._temp_diag_updater = DiagnosticUpdater(
+            '/ros_system_monitor/{}/cpu/temp'.format(namespace)
```
I think I would just make one DiagnosticUpdater for the CPU. The temp and the usage can be their own individual GenericDiagnostic components that both feed into the one CPU status.

There's not really a need for an OutputDiagnostic, because this node doesn't publish anything. That's really designed to be used for monitoring publish frequencies and nothing depends on the specific timing of the CPU checks. If you just want to make sure the event loop that updates the CPU stats doesn't, for some reason stop working, you could do a really simple watchdog thread that just checks the last time it was updated or something. Much lighter weight than all the frequency monitoring logic wrapped up in the OutputDiagnostic. But even the watchdog thread I would say might be overkill, but up to you.

As a separate comment, we should try to standardize on ROS namespacing. We can actually just grab the namespace of this node directly from ROS and prepend it to the stat name. So it would be like `"{}/cpu".format(<get ros namespace>)`

These same comments apply to all the files in this PR.

abrahambotros replied:
> I think I would just make one DiagnosticUpdater for the CPU. The temp and the usage can be their own individual GenericDiagnostic components that both feed into the one CPU status.

Done in https://github.com/MarbleInc/ros-system-monitor/pull/3/commits/e53e25608e1995bfb3f4300ff859f305b691c337. This is what I initially had at first, but changed since RSP was sort of treating these as two different inputs, and since having two separate updaters makes it easier to distinguish which of the two root causes is creating a generic error message. But cool, happy to simplify it to a single updater for each monitor. Applied to both the CPU and HDD monitors.

> There's not really a need for an OutputDiagnostic

Done in https://github.com/MarbleInc/ros-system-monitor/pull/3/commits/9614dde4c488ba6a608df96ee687d19da168f562. Just removed these entirely in all monitors. Our mbot diagnostics watchdog is watching this anyway, I just initially thought it was easy and nice to have. Yeah, I was re-using the machinery of the `OutputDiagnostic` just to easily monitor something that was operating kind of like a publisher, but good point that a simple watchdog thread would be much more light weight even if a bit more code here.

> As a separate comment, we should try to standardize on ROS namespacing

That's kinda what they're already doing in the main function at the bottom of this file already. That returns just `com-1` or `com-2`. Looks like in mbot we pass the namespace manually in the launch file around here: https://github.com/MarbleInc/mbot/blob/c90a2bf75bbd7509a427f4b8ffaed87cf38bf3b9/src/ros/mbot_bringup/drivers/drivers_burrata.launch#L86 . If we wanted to switch to just `{}/cpu.format(namespace)` here, we'd have to edit it in `drivers_burrata.launch` to something like `ns="system_monitor/com-1"` here (and maybe rename `timing_monitor.launch` to `ns="timing_monitor"`...). Thoughts?

zacwitte replied:
Cool, first two points look good.

For namespacing, I think the ros_system_monitor part of the namespace should also come from the launch file. You can have namespaces inside of namespaces. In the diagnostic, the namespace should always be prepended, not inserted into the middle. That means we'll probably end up with `/com-1/ros_system_monitor/cpu` which is a change but that's ok. It makes more sense that way and it's consistent with how the rest of ROS namespacing works.

abrahambotros replied:
Gotcha, thanks for the notes on namespacing. Updated here in https://github.com/MarbleInc/ros-system-monitor/pull/3/commits/330f7b9ad0c8d7503ccfac420cc74e7a8b0fa662; corresponding mbot updates in https://github.com/MarbleInc/mbot/pull/3196/commits/2eb52b5cca9dffff98cb9b81b4c5c053108374bc as part of PR https://github.com/MarbleInc/mbot/pull/3196. More info in commit messages.

Hopefully this aligns well with what you were advising, but let me know if not, thanks!

zacwitte commented on bin/cpu_monitor.py:
```diff
@@ -512,14 +529,21 @@ def publish_stats(self):
-                self._diag_pub.publish(msg)
-                self._last_publish_time = rospy.get_time()
+            # Convert from ROS diagnostics to mbot_diagnostics for publishing.
+            self._temp_diagnostic.set_status(
```
It looks like this is creating a D1.0 status message, updating the properties of that message, then when publishing, copying those properties back to the 2.0 GenericDiagnostic. Why not update just update the GenericDiagnostic component directly and get rid of the D1.0 intermediary?

This comment applies to all the files in this PR.

abrahambotros replied:
Correct, instead of making pretty heavy-handed updates to the internals of many monitor functions, I thought it might be reasonable to just change the interface to the rest of the system - internal work in 1.0, publishing via 2.0. A bit of a quick-and-dirty compromise. This being forked code that does not necessarily need to be uniform with our codebase definitely played a critical role in that decision. What are your thoughts on that, before I do any re-writing?

zacwitte replied:
Yeah it is forked so if we were trying to keep it current with upstream commits I could see the motivation for minimizing changes. I think we've made enough changes to it that we probably won't be pulling in updates. That said, what you have works so it's fine to keep as-is, just an opportunity to simplify. Up to you.

abrahambotros replied:
Yeah, agreed that it's a partially moot point since we're not trying to pull in/keep up-to-date with upstream. It is kinda nice to see a minimal list of changes from upstream (assuming that's stable) if we're ever trying to track down tricky bugs in the future that we may have introduced. But overall mainly just trying to keep the number of changes we make to known/assumed stable code minimal, especially if additional changes are not mission-critical.

Anyway, thanks for discussing this and marking resolved.

Commit:
Combine separate DiagnosticUpdaters

Commit:
Remove publish/output diagnostics on all monitors

Commit:
Rely on namespace for diagnostic topics; remove namespace from messages

- Mbot launch file namespaces all nodes appropriately (ex:
  `/com-X/system_monitor`). Rely on this namespace when naming diagnostic
  topics we publish on.
- Since mbot diagnostic messages are labeled by the diagnostic topic
  (which includes the machine/namespace), remove the old machine/namespace
  usage from each message to keep it concise and less redundant.

Commit:
Set DiagnosticUpdater display_names using diag-hostname param

Use diag-hostname CLI param to construct display_name to use in
each node's DiagnosticUpdater (mbot Diagnostics 2.0).

abrahambotros commented:
@zacwitte https://github.com/MarbleInc/ros-system-monitor/pull/3/commits/4efe217811ca6c0dce5599f3925c2f9c34e6ea80 sets the Diagnostics 2.0 `display_name` in each DiagnosticUpdater. We don't have easy access to the machine name in these scripts from the mbot launch file, so instead pass it in as the diag-hostname CLI param in the mbot launch file (see https://github.com/MarbleInc/mbot/pull/3196/files#diff-73ed6320bd3565b5d9551654561f4d74).

Also, I verified that the namespace available in each node here indeed reliably has a trailing slash (after the namespace set by the parent `<group />` element). I believe this is expected behavior for namespaces set on parent elements of a ROS node in a launch file; https://answers.ros.org/question/314560/namespaces-of-nodes/?answer=314594#post-id-314594 also mentions similar. Since this seems consistent, I'd say we can hold off on introducing some logic to manually handle adding a trailing slash if missing in each of these scripts, but let me know if you still think that should still be implemented here.
